pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,445 @@
+		return false
+	}
+
+	if podIP == "" || podNodeIP == "" || pod.Spec.HostNetwork {
+		klog.V(2).InfoS("Pod has no valid IPs or it is in HostNetwork, skip it", "name", klog.KObj(pod),
+			"podIP", podIP, "nodeIP", podNodeIP, "hostNetwork", pod.Spec.HostNetwork)
+		return false
+	}
+
+	ips, exists := c.getPodNodeIPs(types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name})
+	if exists && ips.podNodeIP == podNodeIP && ips.podIP == podIP {
+		klog.V(2).InfoS("No Pod change impacts installed Openflow rules", "name", klog.KObj(pod))
+		return false
+	}
+	return true
+}
+
+func (c *MCPodRouteController) enqueuePod(obj interface{}, isDelete bool) {
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		klog.ErrorS(errors.New("received unexpected object"), "enqueuePod can't process event", "obj", obj)
+		return
+	}
+
+	c.podQueue.Add(types.NamespacedName{
+		Namespace: pod.Namespace,
+		Name:      pod.Name,
+	})
+}
+
+func (c *MCPodRouteController) Run(stopCh <-chan struct{}) {
+	defer c.gwQueue.ShutDown()
+	defer c.podQueue.ShutDown()
+
+	klog.InfoS("Starting controller", "controller", podRouteControllerName)
+	defer klog.InfoS("Shutting down controller", "controller", podRouteControllerName)
+	if !cache.WaitForNamedCacheSync(podRouteControllerName, stopCh, c.gwInformer.HasSynced) {
+		return
+	}
+
+	for i := 0; i < defaultWorkerNum; i++ {
+		go wait.Until(c.gatewayWorker, time.Second, stopCh)
+	}
+	go handlePodFlowController(c)
+	<-stopCh
+}
+
+func (c *MCPodRouteController) gatewayWorker() {
+	for c.processGatewayNextWorkItem() {
+	}
+}
+
+func (c *MCPodRouteController) processGatewayNextWorkItem() bool {
+	key, quit := c.gwQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.gwQueue.Done(key)
+
+	if k, ok := key.(string); !ok {
+		c.gwQueue.Forget(k)
+		klog.InfoS("Expected string in work queue but got", "object", k)
+		return true
+	} else if err := c.syncGateway(k); err == nil {
+		c.gwQueue.Forget(key)
+	} else {
+		c.gwQueue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing Gateway, requeuing", "key", key)
+	}
+	return true
+}
+
+func (c *MCPodRouteController) isStarted() bool {
+	return c.started.Load().(bool)
+}
+
+func (c *MCPodRouteController) setStarted(started bool) {
+	c.started.Store(started)
+}
+
+func (c *MCPodRouteController) syncGateway(key string) error {
+	activeGW, err := getActiveGateway(c.gwLister, c.namespace)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get an active Gateway")
+		return err
+	}
+
+	// Stop Pod flow controller and clean up all installed Multi-cluster Pod flows
+	// If the Node was a Gateway before.
+	if activeGW == nil || (activeGW != nil && c.nodeConfig.Name != activeGW.Name) {
+		if c.isStarted() {
+			c.finishCh <- struct{}{}
+		}
+		if len(c.podInfos) > 0 {
+			return c.deletePodFlows()
+		}
+		return nil
+	}
+
+	if c.nodeConfig.Name == activeGW.Name {
+		if !c.isStarted() {
+			c.startCh <- struct{}{}
+			return nil
+		} else {
+			return c.installPodFlows()
```

Comment:
Pod controller after starts will anyway process every Pod so this is not necessary, unless you had optimizations to handle all Pods in a batch in `installPodFlows()` but I did not see that.

Reply:
Yeah, removed this.